qemu/tests/ksm_ksmtuned.py

```diff
@@ -0,0 +1,104 @@
+    4. Get the free memory in host
+    5. If both the free memory size is not smaller than the threshold and guest
+        used memory + threshold is not bigger than total memory in host. Check
+        the ksm status in host. Ksm should not start in the host
+    6. Repeat step 2~5 under it broke the rule in step 5
+
+    :param test: kvm test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    def check_ksm(mem, stress=False):
+        """
+        :param mem: Boot guest with given memory, in KB
+        :param stress: Load stress or not
+        """
+        params['mem'] = mem // 1024
+        params['start_vm'] = 'yes'
+        vm_name = params['main_vm']
+        env_process.preprocess_vm(test, params, env, vm_name)
+        vm = env.get_vm(vm_name)
+        vm.wait_for_login()
+        if stress:
+            params['stress_args'] = ('--cpu 4 --io 4 --vm 2 --vm-bytes %sM' %
+                                     (int(params['mem']) // 2))
+            stress_test = VMStress(vm, "stress", params)
+            stress_test.load_stress_tool()
+            time.sleep(30)
+        qemu_pid = vm.get_pid()
+        qemu_used_page = utils_misc.normalize_data_size(process.getoutput(
+            params['cmd_get_qemu_used_mem'] % qemu_pid, shell=True) + 'K', 'B')
+        pagesize = utils_memory.getpagesize()
+        qemu_used_mem = int(float(qemu_used_page)) * pagesize
+        free_mem_host = utils_memory.freememtotal()
+        ksm_status = process.getoutput(params['cmd_check_ksm_status'])
+        vm.destroy()
+        logging.info('The ksm threshold is %s, the memory allocated by qemu is'
+                     ' %s, and the total free memory on host is %s.' %
+                     (ksm_thres, qemu_used_mem, free_mem_host))
+        if free_mem_host >= ksm_thres:
+            if ksm_status != '0':
+                test.fail('Ksm should not start.')
+            if stress:
+                test.error('The host resource is not consumed as expected.')
+        elif ksm_status == '0':
+            test.fail('Ksm should start but it does not.')
+
+    total_mem_host = utils_memory.memtotal()
+    utils_memory.drop_caches()
+    free_mem_host = utils_memory.freememtotal()
+    ksm_thres = process.getoutput(params['cmd_get_thres'], shell=True)
+    ksm_thres = int(total_mem_host *
+                    (int(re.findall('\\d+', ksm_thres)[0]) / 100))
+    guest_mem = (free_mem_host - ksm_thres) // 2
+    if arch.ARCH in ('ppc64', 'ppc64le'):
+        guest_mem = guest_mem - guest_mem % (256 * 1024)
+    status_ksm_service = process.system(
+        params['cmd_status_ksmtuned'], ignore_status=True)
+    if status_ksm_service != 0:
+        process.run(params['cmd_start_ksmtuned'])
+    check_ksm(guest_mem)
+
+    ksm_config_file = params['ksm_config_file']
+    backup_file = ksm_config_file + '.backup'
+    copyfile(ksm_config_file, backup_file)
+    threshold = 60
```

Comment:
Can we incease threshold a bit for ppc, the service will be triggered while threshold is 70 and 80, thanks.

Reply:
Prefer to use a parameter for threshold for different architectures. Thanks

Reply:
@MiriamDeng Yes, updated, please help check, thanks.